### DISCUSSION TITLE
Can Arrow be used to create an annotation to enforce parameters in DSLs?

### DISCUSSION THREAD

**Discussion by elect86** [arrow-kt/arrow#2735] [General]
As titled, I was reading this [comment](https://www.reddit.com/r/Kotlin/comments/if2hgj/comment/g2mfoey/?utm_source=share&utm_medium=web2x&context=3) and he mentioned Arrow (meta)

would this be possible?

Ps: I'm looking for something like this:
```
@DSLint
class Person {

    @set:DSLMandatory
    var name: String
}
```

![image](https://miro.medium.com/max/1400/1*lzanVUqZD7kfnX3cuiBNOQ.png)

**Comment by raulraja** [comment]
Hi @elect86, 
Yes, this would be done with a `callChecker { }` where you look in the body for missing initialized vars from Person.
The problem is the current call checker is for the legacy compiler API as FIR (Frontend IR) is still in development by Jetbrains. I believe in 1.7.x o 1.8 we will have a call checker for FIR in Meta.

In the meantime, you may use the legacy API since call checkers, expression checkers, and declaration checkers are very similar to how they are in FIR and won't be hard to migrate when available.

An example of a call checker can be found here https://github.com/arrow-kt/arrow-meta/blob/56a5b55415fe697026c7ede3a4d35813e02fe1e1/compiler-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/resolve/ProofResolutionRules.kt#L82

Feel free to ask any questions you have during development in the kotlin slack #arrow-meta channel.

**Comment by kyay10** [comment]
Arrow Meta is K1-only, so I'm closing this as outdated.
For K2, have a look at FIR compiler plugins. These [docs](https://github.com/JetBrains/kotlin/tree/master/docs/fir) should help
